README.md
# four.js

An attemp to render 4 dimensional objects with three.js.

**Full demo: [anakata](https://florian.mounier.dev/anakataGL/)**

## Installation

```bash
yarn add four-js
# or
npm --save install four-js
```

## Usage

Full working example [here](https://github.com/paradoxxxzero/four.js/blob/master/examples/tesseract.html)

To run it: `git clone https://github.com/paradoxxxzero/four.js && cd four.js && yarn install && yarn start` then open http://localhost:44444/examples/tesseract.html

### Creating

```js
import { HyperGeometry, shapes } from 'four-js'

// Import tesseract preset shape
const { tesseract } = shapes

// Create an HyperRenderer which will be used to make 4d -> 3d projections
const hyperRenderer = new HyperRenderer(1.5, 5)
```

#### 4D -> 3D Projection

```js
// Instantiate an HyperMesh with the shape
const hyperMesh = new HyperMesh(shape)

// (...) Setup the three.js scene as usual and add the HyperMesh:
scene.add(hyperMesh)
```

#### 4D -> 3D Cross Section

```js
// Instantiate an HyperSlice with the shape
const hyperSlice = new HyperSlice(shape)

// (...) Setup the three.js scene as usual and add the HyperSlice:
scene.add(hyperSlice)
```

### Updating

This will render a tesseract that you can then rotate by updating the HyperRenderer rotation:

```js
update() {
  requestAnimationFrame(update)
  // Rotate takes the rotation speed around the 6 planes:
  hyperRenderer.rotate({ xy: 0, xz: 0, xw: 5, yz: 0, yw: 10, zw: 10 })

  // Move the cross section along the w-axis
  hyperRenderer.shiftSlice(0.5, wmin, wmax)

  // Update the hyperMesh
  hyperMesh.update(hyperRenderer)

  // Update the hyperSlice
  hyperSlice.update(hyperRenderer)

  // Render the scene
  renderer.render(scene, camera)
}
```

### Configuration

`HyperMesh` and `HyperSlice` accept a configuration object as second argument:

```js
const meshConfig = {
  faces: {
    // Configuration regarding faces
    enabled: true, // Render faces
    useColors: true, // Use vertex coloring through colorGenerator
    colorGenerator: cellColors, // Called with ({shape, colors}) must return a function that returns a colors from ({cell, face, vertex, type}) indexes
    colors: defaultColors, // Colors to chose from
    reuse: 'none', // One of ['all', 'faces', 'none'], specifies whether to duplicate vertex in faces / cells. Useful for polychora due to needing a normal per face per vertex.
    split: 'cells', // One of ['none', 'cells', 'faces'], specifies whether to render one Mesh, one Mesh per cell, one Mesh pes face
    splitScale: 100, // Scale of each split, allow for better comprehension of 4D Mesh when used with split: 'cells'
    material: new MeshPhongMaterial({
      // The material (or list of materials per split) to use when rendering
      transparent: true,
      opacity: 0.25,
      blending: NormalBlending,
      depthWrite: false,
      side: DoubleSide,
      vertexColors: true,
    }),
  },
  edges: {
    // Render edges
    enabled: true,
    useColors: true,
    colorGenerator: cellColors,
    colors: defaultColors,
    reuse: 'faces',
    split: 'cells',
    splitScale: 100,
    material: new LineBasicMaterial({
      transparent: true,
      opacity: 0.25,
      blending: AdditiveBlending,
      depthWrite: false,
      vertexColors: true,
      linewidth: 2,
    }),
  },
  points: {
    // Render points
    enabled: false,
    useColors: true,
    colorGenerator: cellColors,
    colors: defaultColors,
    reuse: 'faces',
    split: 'none',
    splitScale: 100,
    material: new ShaderMaterial({
      uniforms: {
        size: { value: 5 },
        opacity: { value: 0.25 },
      },
      vertexShader: pointsVertexShader,
      fragmentShader: pointsFragmentShader,
      transparent: true,
      blending: AdditiveBlending,
    }),
  },
}

const hyperMesh = new HyperMesh(shape, meshConfig)

// Same configuration (minus split/reuse configs) for HyperSlice:
const sliceConfig = {
  faces: {
    enabled: true,
    useColors: true,
    colorGenerator: cellColors,
    colors: defaultColors,
    material: new MeshPhongMaterial({
      side: DoubleSide,
      shininess: 50,
      vertexColors: true,
    }),
  },
  edges: {
    enabled: true,
    useColors: true,
    colorGenerator: cellColors,
    colors: defaultColors,
    material: new LineBasicMaterial({
      transparent: true,
      opacity: 0.25,
      blending: AdditiveBlending,
      depthWrite: false,
      vertexColors: true,
      linewidth: 2,
    }),
  },
  points: {
    enabled: false,
    useColors: true,
    colorGenerator: cellColors,
    colors: defaultColors,
    material: new ShaderMaterial({
      uniforms: {
        size: { value: 5 },
        opacity: { value: 0.25 },
      },
      vertexShader: pointsVertexShader,
      fragmentShader: pointsFragmentShader,
      transparent: true,
      blending: AdditiveBlending,
    }),
  },
}

const hyperSlice = new HyperSlice(shape, sliceConfig)
```

## Shape definition

The 4 dimensional mesh descriptions are composed of vertices, faces and cells.

Here's what the tesseract description look like:

```js
const tesseract = {
  // These are the classic vertices
  vertices: [
    [1, 1, 1, 1], // 0
    [1, 1, -1, 1], // 1
    [1, -1, -1, 1], // 2
    [1, -1, 1, 1], // 3
    [-1, 1, 1, 1], // 4
    [-1, 1, -1, 1], // 5
    [-1, -1, -1, 1], // 6
    [-1, -1, 1, 1], // 7
    [1, 1, 1, -1], // 8
    [1, 1, -1, -1], // 9
    [1, -1, -1, -1], // 10
    [1, -1, 1, -1], // 11
    [-1, 1, 1, -1], // 12
    [-1, 1, -1, -1], // 13
    [-1, -1, -1, -1], // 14
    [-1, -1, 1, -1], // 15
  ],
  // Each face is made by listing its vertices index
  faces: [
    [0, 1, 2, 3], // 0
    [0, 4, 5, 1], // 1
    [0, 3, 7, 4], // 2
    [3, 2, 6, 7], // 3
    [1, 5, 6, 2], // 4
    [4, 7, 6, 5], // 5
    [0, 1, 9, 8], // 6
    [4, 5, 13, 12], // 7
    [3, 2, 10, 11], // 8
    [7, 6, 14, 15], // 9
    [0, 3, 11, 8], // 10
    [4, 7, 15, 12], // 11
    [1, 2, 10, 9], // 12
    [5, 6, 14, 13], // 13
    [0, 4, 12, 8], // 14
    [1, 5, 13, 9], // 15
    [2, 6, 14, 10], // 16
    [3, 7, 15, 11], // 17
    [11, 10, 9, 8], // 18
    [9, 13, 12, 8], // 19
    [12, 15, 11, 8], // 20
    [15, 14, 10, 11], // 21
    [10, 14, 13, 9], // 22
    [13, 14, 15, 12], // 23
  ],
  // Each cell is made by listing its faces index
  cells: [
    [0, 1, 2, 3, 4, 5], // 0
    [0, 6, 12, 8, 10, 18], // 1
    [1, 6, 14, 7, 15, 19], // 2
    [4, 12, 16, 13, 15, 22], // 3
    [3, 8, 16, 9, 17, 21], // 4
    [2, 10, 17, 11, 14, 20], // 5
    [5, 7, 13, 9, 11, 23], // 6
    [18, 19, 20, 21, 22, 23], // 7
  ],
}
```

This means that the first cell

```js
    [0, 1, 2, 3, 4, 5], // 0
```

is composed of 6 faces:

```js
    [0, 1, 2, 3], // 0
    [0, 4, 5, 1], // 1
    [0, 3, 7, 4], // 2
    [3, 2, 6, 7], // 3
    [1, 5, 6, 2], // 4
    [4, 7, 6, 5], // 5
```

and the first face here

```js
    [0, 1, 2, 3], // 0
```

is composed of 4 vertices:

```js
    [1, 1, 1, 1], // 0
    [1, 1, -1, 1], // 1
    [1, -1, -1, 1], // 2
    [1, -1, 1, 1], // 3
```

Following that logic you can try to draw your own 4d models.

Feel free to make pull requests with your own creations!

# Contributors

### @FranzPoize

For his nice 3-sphere shape which has since been refactored and generalized into `uvw-hypersurfaces`.


CHANGELOG.md
# Changelog

## [Unreleased]

## [2.1.0] - 2022-07-07

### Changed

- Use `StreamDrawUsage` instead of `DynamicDrawUsage` to increase performance.
- Update dependencies

## [2.0.1] - 2021-06-16

## Added

- `HyperRendererCached` which is a rotation cached version of `HyperRenderer` a noticeable optimization for `HyperSlice`.

## Fixed

- Crashes in colorGenerator when there's only one color and with depth on slice.

# [2.0.0] - 2021-06-13

### Breaking Changes

- API complete overhaul, all hyper geometries has been merged into `HyperMesh` see [README.md] for more information.

### Added

- `HyperSlice` to visualize shapes with 3-plane cross sections.
- `normalizeShape` helper to constraint shape in [-1, 1] size.
- `colorGenerator` module with some default generators for use with material vertexColors.

### Removed

- `HyperGeometry`, `HyperEdgesGeometry`, `HyperPointsGeometry`, `HyperGeometryMergedVertices` and `HyperEdgesGeometryMergedEdges` (merged into `HyperMesh`)

## [1.1.0] - 2021-06-01

### Breaking Changes

- Renamed `HyperEdgeGeometry` to `HyperEdgesGeometry`
- Changed signature of `HyperEdgesGeometry` and `HyperPointsGeometry` to accept an `HyperGeometry` instead of shape `vertices`, `faces`, `cells` to avoid multiple projections. The `HyperGeometry` must now be updated even if not rendered.

### Added

- `HyperGeometryMergedVertices` which further reuse vertices at the expense of having one normal by vertex.

- `HyperEdgesGeometryMergedEdges` which use only one Line for performance (but doesn't support cellSize).

## [1.0.3] - 2021-05-31

### Added

- Add remaining sage shapes:
  - `hecatonicosachoronRectified`
  - `hecatonicosachoronRuncinated`
  - `hecatonicosachoronRuncitruncated`
  - `hecatonicosachoronOmnitruncated`
  - `hecatonicosachoronTruncated`
  - `hexacosichoronRectified`
  - `hexacosichoronRuncitruncated`
  - `hexacosichoronOmnitruncated`
  - `hexacosichoronTruncated`
  - `permutahedronH`

## [1.0.2] - 2021-05-23

### Added

- `uLoop`, `vLoop` parameters for `generateUVSurface` to control if the surface should close on itself
- `uLoop`, `vLoop`, `wLoop` parameters for `generateUVWHyperSurface` to control if the hypersurface should close on itself

- `hecatonicosachoronCantitruncated` and `hexacosichoronCantitruncated` shapes

## [1.0.1] - 2021-05-22

- Rerelease (with correct built version...)

## [1.0.0] - 2021-05-22

### Breaking Changes

- Some shapes has been renamed:
  - `threesphere` -> `glome`
  - `threetorus` -> `diTorus`
  - `flattorus` -> `flatTorus`

### Added

- Add generators for simple polytopes to generate custom sizes: `generateTesseract`, `generateHyperCuboid`, `generateCuboidPyramid`, `generateCubicPyramid`, `generateGlome`, `generateDiTorus`, `generateFlatTorus`,`generateDuoCylinder`
- Add uv surface generator and uvw hypersurface generator: `generateUVSurface`, `generateUVWHyperSurface`
- Add lot of shapes:

  - `hyperCuboid`
  - `cubicPyramid`
  - `cubicBiPyramid`
  - `cuboidPyramid`
  - `cuboidBiPyramid`
  - `octahedralPyramid`
  - `octahedralBiPyramid`
  - `octahedraloidPyramid`
  - `octahedraloidBiPyramid`
  - `grandAntiPrism`
  - `hecatonicosachoron`
  - `hecatonicosachoronCantellated`
  - `hexacosichoron`
  - `hexacosichoronBitTruncated`
  - `hexacosichoronCantellated`
  - `icositetrachoron`
  - `buckyBallPyramid`
  - `minkowskiSumOfTesseractAndHexadecachoron`
  - `permutahedronA`
  - `permutahedronAIrregular`
  - `permutahedronB`
  - `permutahedronBIrregular`
  - `permutahedronC`
  - `permutahedronCIrregular`
  - `permutahedronD`
  - `permutahedronDIrregular`
  - `permutahedronF`
  - `permutahedronFIrregular`
  - `permutahedronH`
  - `permutahedronHIrregular`
  - `crossHyperSurface`
  - `rightHyperCone`
  - `obliqueHyperCone`
  - `spherinder`
  - `torinder`
  - `duoCylinder`
  - `cubinder`
  - `tiger`
  - `torisphere`
  - `spheritorus`

- Glome and DiTorus are now generated with generateUVWHyperSurface

## [0.1.3] - 2021-05-15

### Added

- Cubic pyramid
- Grand Antiprism
- Hecatonicosachoron (120-cell)
- Hexacosichoron (600-cell)
- Icositetrachoron (24-cell)

## [0.1.2] - 2021-03-01

### Added

- 3-torus and flat-torus shapes thanks to @FranzPoise via #2

## [0.1.1] - 2021-02-11

### Added

- 3-sphere shape thanks to @FranzPoise via #1

## [0.1.0] - 2021-02-03

### Added

- All rendering components: `HyperMesh`, `HyperRenderer`, `HyperGeometry` `HyperEdgeGeometry`, `HyperPointsGeometry` and 3 shapes: `tesseract`, `pentachoron`, `hexadecachoron`.

[unreleased]: https://github.com/paradoxxxzero/four.js/compare/v2.1.0...HEAD
[2.0.1]: https://github.com/paradoxxxzero/four.js/compare/v2.0.1...v2.1.0
[2.0.1]: https://github.com/paradoxxxzero/four.js/compare/v2.0.0...v2.0.1
[2.0.0]: https://github.com/paradoxxxzero/four.js/compare/v1.1.0...v2.0.0
[1.1.0]: https://github.com/paradoxxxzero/four.js/compare/v1.0.3...v1.1.0
[1.0.3]: https://github.com/paradoxxxzero/four.js/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/paradoxxxzero/four.js/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/paradoxxxzero/four.js/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/paradoxxxzero/four.js/compare/v0.1.3...v1.0.0
[0.1.3]: https://github.com/paradoxxxzero/four.js/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/paradoxxxzero/four.js/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/paradoxxxzero/four.js/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/paradoxxxzero/four.js/compare/...v0.1.0


src/shapes.js
export const tesseract = {
  vertices: [
    [1, 1, 1, 1], // 0
    [1, 1, -1, 1], // 1
    [1, -1, -1, 1], // 2
    [1, -1, 1, 1], // 3
    [-1, 1, 1, 1], // 4
    [-1, 1, -1, 1], // 5
    [-1, -1, -1, 1], // 6
    [-1, -1, 1, 1], // 7
    [1, 1, 1, -1], // 8
    [1, 1, -1, -1], // 9
    [1, -1, -1, -1], // 10
    [1, -1, 1, -1], // 11
    [-1, 1, 1, -1], // 12
    [-1, 1, -1, -1], // 13
    [-1, -1, -1, -1], // 14
    [-1, -1, 1, -1], // 15
  ],
  faces: [
    [0, 1, 2, 3], // 0
    [0, 4, 5, 1], // 1
    [0, 3, 7, 4], // 2
    [3, 2, 6, 7], // 3
    [1, 5, 6, 2], // 4
    [4, 7, 6, 5], // 5

    [0, 1, 9, 8], // 6
    [4, 5, 13, 12], // 7
    [3, 2, 10, 11], // 8
    [7, 6, 14, 15], // 9

    [0, 3, 11, 8], // 10
    [4, 7, 15, 12], // 11
    [1, 2, 10, 9], // 12
    [5, 6, 14, 13], // 13

    [0, 4, 12, 8], // 14
    [1, 5, 13, 9], // 15
    [2, 6, 14, 10], // 16
    [3, 7, 15, 11], // 17

    [11, 10, 9, 8], // 18
    [9, 13, 12, 8], // 19
    [12, 15, 11, 8], // 20
    [15, 14, 10, 11], // 21
    [10, 14, 13, 9], // 22
    [13, 14, 15, 12], // 23
  ],
  cells: [
    [0, 1, 2, 3, 4, 5], // 0
    [0, 6, 12, 8, 10, 18], // 1
    [1, 6, 14, 7, 15, 19], // 2
    [4, 12, 16, 13, 15, 22], // 3
    [3, 8, 16, 9, 17, 21], // 4
    [2, 10, 17, 11, 14, 20], // 5
    [5, 7, 13, 9, 11, 23], // 6
    [18, 19, 20, 21, 22, 23], // 7
  ],
}

export const pentachoron = {
  vertices: [
    [1, 1, 1, -1 / Math.sqrt(5)], // 0
    [1, -1, -1, -1 / Math.sqrt(5)], // 1
    [-1, 1, -1, -1 / Math.sqrt(5)], // 2
    [-1, -1, 1, -1 / Math.sqrt(5)], // 3
    [0, 0, 0, Math.sqrt(5) - 1 / Math.sqrt(5)], // 4
  ],
  faces: [
    [1, 2, 3], // 0
    [0, 1, 2], // 1
    [0, 1, 3], // 2
    [0, 3, 2], // 3

    [0, 4, 1], // 4
    [0, 2, 4], // 5
    [0, 3, 4], // 6

    [2, 4, 3], // 7
    [1, 3, 4], // 8
    [1, 4, 2], // 9
  ],
  cells: [
    [0, 1, 2, 3], // 0
    [1, 5, 4, 9], // 1
    [3, 6, 5, 7], // 2
    [2, 4, 6, 8], // 3
    [0, 7, 8, 9], // 4
  ],
}

export const hexadecachoron = {
  vertices: [
    [-1, 0, 0, 0], // 0
    [1, 0, 0, 0], // 1
    [0, -1, 0, 0], // 2
    [0, 1, 0, 0], // 3
    [0, 0, -1, 0], // 4
    [0, 0, 1, 0], // 5
    [-0, 0, 0, -1], // 6
    [0, 0, 0, 1], // 7
  ],
  faces: [
    [0, 2, 4], // 0
    [0, 2, 5], // 1
    [0, 2, 6], // 2
    [0, 2, 7], // 3
    [0, 3, 4], // 4
    [0, 3, 5], // 5
    [0, 3, 6], // 6
    [0, 3, 7], // 7
    [0, 4, 6], // 8
    [0, 4, 7], // 9
    [0, 5, 6], // 10
    [0, 5, 7], // 11
    [1, 2, 4], // 12
    [1, 2, 5], // 13
    [1, 2, 6], // 14
    [1, 2, 7], // 15
    [1, 3, 4], // 16
    [1, 3, 5], // 17
    [1, 3, 6], // 18
    [1, 3, 7], // 19
    [1, 4, 6], // 20
    [1, 4, 7], // 21
    [1, 5, 6], // 22
    [1, 5, 7], // 23
    [2, 4, 6], // 24
    [2, 4, 7], // 25
    [2, 5, 6], // 26
    [2, 5, 7], // 27
    [3, 4, 6], // 28
    [3, 4, 7], // 29
    [3, 5, 6], // 30
    [3, 5, 7], // 31
  ],
  cells: [
    [0, 2, 8, 24], // 0  v: 0, 2, 4, 6
    [0, 3, 9, 25], // 1  v: 0, 2, 4, 7
    [1, 2, 10, 26], // 2  v: 0, 2, 5, 6
    [1, 3, 11, 27], // 3  v: 0, 2, 5, 7
    [4, 6, 8, 28], // 4  v: 0, 3, 4, 6
    [4, 7, 9, 29], // 5  v: 0, 3, 4, 7
    [5, 6, 10, 30], // 6  v: 0, 3, 5, 6
    [5, 7, 11, 31], // 7  v: 0, 3, 5, 7

    [12, 14, 20, 24], // 8  v: 1, 2, 4, 6
    [12, 15, 21, 25], // 9  v: 1, 2, 4, 7
    [13, 14, 22, 26], // 10  v: 1, 2, 5, 6
    [13, 15, 23, 27], // 11  v: 1, 2, 5, 7
    [16, 18, 20, 28], // 12  v: 1, 3, 4, 6
    [16, 19, 21, 29], // 13  v: 1, 3, 4, 7
    [17, 18, 22, 30], // 14  v: 1, 3, 5, 6
    [17, 19, 23, 31], // 15  v: 1, 3, 5, 7
  ],
}
